module/VuFindSearch/src/VuFindSearch/Backend/Solr/QueryBuilder.php

```diff
@@ -210,6 +212,64 @@ public function build(AbstractQuery $query, ?ParamBag $params = null)
         return $newParams;
     }
 
+    /**
+     * Converts a simple query (Query) into an advanced one (QueryGroup) if part of it should be an exact query.
+     * This only supports a single exact query (surrounded with quotes) combined with a non-exact query.
+     * Logical operators can be used, but not parenthesis or field names.
+     * The original query is returned for any non-supported case.
+     *
+     * @param QueryGroup|Query $query User query
+     *
+     * @return QueryGroup|Query
+     */
+    protected function possiblyConvertMixedExactQueryIntoAdvanced($query)
+    {
+        if ($query instanceof QueryGroup) {
+            return $query;
+        }
```

Comment:
Would be nice to use QueryInterface for greater flexibility, and have more explicit types. Maybe something like:

```suggestion
     * @param QueryInterface $query User query
     *
     * @return QueryInterface
     */
    protected function possiblyConvertMixedExactQueryIntoAdvanced(QueryInterface $query): QueryInterface
    {
        if (!($query instanceof Query)) {
            return $query;
        }
```

Reply:
Initially I was using `AbstractQuery`, which is used in the calling code and implements `QueryInterface`. But my editor (with the intelephense extension) was complaining that it doesn't define `getHandler()` (which is only in `Query`). I tried for a while to do a cast and use a different variable with the `Query` type specified in an inline comment. It worked but it was ugly. Then I thought of using `QueryGroup|Query`. The type checker is smart enough to realize that if the initial type is `QueryGroup|Query` and if `QueryGroup` objects are discarded, it only leaves the `Query` type, and it no longer complains about the method call. I like this solution better, because it clarifies the expected types in the function definition.
Actually, now that I think about it, I have a better option: use `QueryInterface` or `AbstractQuery`, and return the query if it's not an instance of `Query`. I am not sure what `WorkKeysQuery` is for, but it would better handle that case. Note that there are similar issues with query types elsewhere in `QueryBuilder`.

Reply:
Agreed, that's why I suggested the `!($query instanceof Query)` check at the top of the method.

It may be worthwhile to invest some effort into cleaning up the other type issues you mention as a dev-12.0 PR (I'm sure problems have accumulated over time as the code has evolved), but I'll leave it to you whether or not now is the time for that. :-)

Reply:
I will go with your solution, just using `AbstractQuery` instead of `QueryInterface` to be more consistent with the rest of the code in the file.

Reply:
Since `AbstractQuery` implements `QueryInterface`, I think using `QueryInterface` would be preferable as it is a more general option. (In practice, I think it likely makes no real difference one way or the other -- but using the interface feels better from a design perspective).

Reply:
That being said, if you think it's better to use `AbstractQuery` for now, and then follow up with a dev-12.0 PR that switches all `AbstractQuery` references to `QueryInterface` for consistency, I could live with that approach. :-)